### PR TITLE
[Feature] FSDP2

### PR DESCRIPTION
**Pull request by sami-bg**
## Description

Adds [PyTorch FSDP2](https://docs.pytorch.org/docs/2.11/distributed.fsdp.fully_shard.html) (`torch.distributed.fsdp.fully_shard`) as a first-class training strategy for stable-pretraining, alongside the existing DDP support. FSDP shards parameters, gradients, and optimizer state across ranks (useful for vit-large+ scale).

The integration is wired through Lightning's `ModelParallelStrategy` (FSDP2). (Lightning's `FSDPStrategy` is FSDP1 and is **incompatible with multi-view SSL methods** that run more than one backbone forward per training step: LeJEPA, DINO, IJEPA, BYOL with multi-view, MAE with multi-block-mask. FSDP1's flat training-state machine asserts on the second forward whereas FSDP2 uses standard PyTorch hooks per call and supports multi-forward by design. This is the principal reason we use FSDP2 specifically.)

This PR also fixes a pre-existing bug in `stable_pretraining/utils/distributed.py` where `all_gather` and `all_reduce` discarded the autograd-aware return values of `torch.distributed.nn.functional.all_gather` / `all_reduce` and always returned the local tensor unchanged.

## How it works

`stable_pretraining/utils/fsdp.py` exposes a Hydra-friendly factory and the helpers needed to make FSDP2 composable into stable-pretraining. The integration point is `Module.configure_model` (the Lightning-named hook `ModelParallelStrategy` requires); it reads the `device_mesh` Lightning sets on the module and dispatches to `default_parallelize_fn`, which:

1. Slices the data-parallel submesh out of Lightning's 2-D `(data_parallel × tensor_parallel)` mesh so `fully_shard` sees exactly the DP axis it expects.
2. Auto-detects transformer / residual block classes (`*Block` / `*Layer` repeated >= 2 times, matches timm ViT, HuggingFace ViT, torchvision ResNet conventions) and applies `fully_shard` per block.
3. Detaches `callbacks_modules` / `callbacks_metrics` containers from the module tree, applies `fully_shard` once at the root, then reattaches them. Skipping callback containers in the per-block loop alone is **not sufficient** — the root sweep would otherwise claim their parameters as DTensors and break each callback's standalone optimizer (OnlineProbe, OnlineKNN, RankMe). Detach/reattach around the root call is the standard FSDP2 way to exclude submodules from the root sweep without giving them their own FSDP unit.
4. Logs what got sharded, what got skipped, and on what mesh, so a misconfigured run is visible in the first few lines of stdout.

`assert_aligned_wrapping` is the DTensor-aware alignment check used by `TeacherStudentWrapper` to guarantee correct in-place EMA under sharding. It checks not just shape and dtype but also `DTensor.placements` and `device_mesh` — two DTensors with matching local shape can still have different placements (`Shard(0)` vs `Shard(1)` on equal-sized dims, or `Shard` vs `Replicate` in degenerate cases), and `zip(teacher.parameters(), student.parameters())` would silently pair non-corresponding shards without that check.

Switching a benchmark from DDP to FSDP2 is a one-line change:

```python
from stable_pretraining.utils.fsdp import make_fsdp_strategy

strategy = make_fsdp_strategy()  # auto-detects data_parallel_size from LOCAL_WORLD_SIZE / cuda.device_count()

trainer = pl.Trainer(
    strategy=strategy,
    accelerator="gpu",
    devices=num_gpus,
    precision="bf16-mixed",  # ModelParallelStrategy doesn't accept "16-mixed";
                             # use bf16-mixed / bf16-true / 16-true / 32-true.
)
```

For custom block selection, mixed-precision policy, or hybrid sharding, subclass `Module` and override `apply_fsdp2_sharding_if_needed`. The `make_fsdp_strategy` factory exposes `mp_policy` directly for common tuning.

## Compatibility notes (gotchas, with the fix in this PR)

- **`Module.validation_step` / `test_step` / `predict_step`** now go through `self(batch, ...)` instead of `self.forward(batch, ...)`. PyTorch's forward pre/post hooks (which is how FSDP2 triggers the all-gather of sharded DTensor params before each fwd) only fire on `__call__`, not `.forward()` call. The previous approach left DTensors ungathered during validation, and the very next op would raise `got mixed torch.Tensor and DTensor`.
- **`Module.named_parameters`** now accepts and forwards `remove_duplicate`. PyTorch's FSDP wrap path always calls `model.named_parameters(remove_duplicate=False)`; the override didn't accept the kwarg, raising `TypeError`
- **`LogUnusedParametersOnce` self-disables under FSDP2**. After `fully_shard`, parameters are DTensors and gradient flow goes through the unsharded gathered tensor that FSDP2 produces inside the forward all-gather. The detector's `Tensor.register_hook` on the DTensor never fires and every parameter would log as "unused" and flood the logs.

## End-to-end validation

FSDP2 was verified end-to-end with a 10-epoch LeJEPA ViT-Small on ImageNet-10 (Imagenette) via `benchmarks/imagenet10/lejepa_vit_small_fsdp.py`, compared against the existing DDP variant and the single-GPU baseline. All three runs use the **same effective batch size (128)** and the **same LR schedule horizon** (`max_epochs=600` cosine, with `StopAfterNEpochs(10)` for the multi-GPU runs to preserve schedule shape). OnlineProbe / OnlineKNN / RankMe trajectories track well as shown below:

image

## Files changed (overview)

- **New**: `stable_pretraining/utils/fsdp.py` — factory, parallelize_fn, alignment check, detection helpers.
- **Modified**: `stable_pretraining/module.py` — `configure_model` hook, `apply_fsdp2_sharding_if_needed`, `named_parameters(remove_duplicate=...)` forwarding, `validation_step` / `test_step` / `predict_step` route through `__call__`.
- **Modified**: `stable_pretraining/manager.py` — FSDP2 strategy detection and logging via `_log_fsdp_info`.
- **Modified**: `stable_pretraining/utils/distributed.py` — `all_gather` / `all_reduce` autograd-aware return-value fix.
- **Modified**: `stable_pretraining/callbacks/unused_parameters.py` self-disables under FSDP2.
- **New**: `benchmarks/imagenet10/lejepa_vit_small_fsdp.py` for the comparison runs.
- **Rename**: dashed benchmark filenames (`lejepa-vit-small.py`, etc.) → underscored, so the variants can `from lejepa_vit_small import ...` directly without `importlib.util` machinery.

## Checklist

- [x] I have read the [**Contributing**](https://rbalestr-lab.github.io/stable-SSL.github.io/dev/contributing.html#) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [ ] I have added the PR to the [**RELEASES.rst**](../RELEASES.rst) file.

### CODE REVIEW

File: benchmarks/imagenet10/lejepa_vit_small_fsdp.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+from lejepa_vit_small import (
+    LeJEPA,
+    _global_transform,
+    _local_transform,
+    lejepa_forward,
+)
+
+
+class StopAfterNEpochs(pl.pytorch.callbacks.Callback):
```
Comment (RandallBalestriero) [inline benchmarks/imagenet10/lejepa_vit_small_fsdp.py]: Do we really need that? cause we have an EarlyStopping callback already

File: stable_pretraining/_version.py
Comment (RandallBalestriero) [inline stable_pretraining/_version.py:1]: this should be ignored right? should be auto generated not pushed from your local?

File: stable_pretraining/utils/fsdp2.py
Comment (RandallBalestriero) [inline stable_pretraining/utils/fsdp2.py:1]: Why do we need that file and not just do `, strategy="fsdp"` into the Trainer?

**Review by RandallBalestriero — CHANGES_REQUESTED** [review]

File: stable_pretraining/utils/fsdp2.py
Comment (sami-bg) [reply stable_pretraining/utils/fsdp2.py:1]: `strategy="fsdp"` uses FSDP1 which, due to its implementation, does not allow anything that isnt just forward -> backward, so doing multiple forwards without a backward in between raises an exception (e.g. with lejepa global->local)

to get around that we need FSDP2 which, with Lightning, is only accessible with `strategy=ModelParallelStrategy(...)`. However! I just found out we can register "fsdp2" with the Lightning [StrategyRegistry](https://lightning.ai/docs/pytorch/stable/advanced/strategy_registry.html), so I think that would be the best approach. This way, users can just use `strategy="fsdp2"`  out of the box, and internally it would use `ModelParallelStrategy(...)`.

Comment (RandallBalestriero) [reply stable_pretraining/utils/fsdp2.py:1]: nice! so we don't need fsdp.py anymore right?

Comment (sami-bg) [reply stable_pretraining/utils/fsdp2.py:1]: i think we still do; it is there as a convenience to enable sharding of timm and hf vit's by default (eg if someone passes a lejepa backbone, it uses a timm vit under the hood, so passing in `strategy="fsdp2"` would just work because it would use `default_parallelize_fn`), and is also a place to house registering the `"fsdp2"` name itself into the strategy registry for lightning. without this, `strategy="fsdp2"` would error out.

**Comment by sami-bg** [comment]
@RandallBalestriero I was working on smoke-testing `fsdp` (built-in) alongside `fsdp2` (what we register for convenience), and `fsdp` seems to fail no matter what, and is kind of unavoidable.

After a bit of debugging it seems the core reason is that we use `manual_backward` in `module.py`, which we did so that we can support multiple optimizers for the probe callbacks in Lightning >= 2.0

The reason has to do with fsdp1's design (and is why things in general are being migrated to `fsdp2` with the `ModelParallelStrategy` in Lighting). In fsdp1, each paramgroup tracks its fsdp-related state with `FORWARD`, `BACKWARD_PRE`, `BACKWARD_POST` and `IDLE`, which let pytorch hook into and know when to all-gather params and when to reduce-scatter gradients across shards. The general lifecycle is:

`IDLE -> FORWARD -> IDLE -> BACKWARD_PRE -> BACKWARD_POST -> IDLE`.

Lightning's FSDPStrategy (i.e. fsdp1) calls our `Module.training_step` via `_forward_redirection`, which is where the `IDLE -> FORWARD -> IDLE` state is managed. This then calls the FSDP1 wrapper's `__call__`. This `__call__` only returns when `training_step` returns, which only returns after `manual_backward` returns. So when we call `manual_backward`, we are technically still in `FORWARD`, and the post-backward hook asserts the state to be `BACKWARD_PRE` or `BACKWARD_POST` and fails. This is not the case with auto-opt as Lightning would run the backward outside the FSDP1 wrapper after `__call__` returns, so the state machine is still consistent, but we cant use `auto-opt` because of multi-optimizer.


fsdp2 sidesteps this completely because it replaces each parameter with a DTensor and then attaches the gather/scatter hooks with the standard `register_forward_pre_hook` and `register_forward_hook`, so it does not redirect and everything remains intact even with our `training_step`.

IMHO I think we should just support fsdp2, as it is equivalent in performance to fsdp1, and the whole ecosystem is moving to fsdp2 regardless. Maybe in practicality we can have a runtime exception against using fsdp to begin with to prevent users from trying, and just redirect them to use fsdp2 instead. It would be just as simple.
